# SQL Part 3 Studio

### Let's practice your SQL quering skills!  For each question, work along in the notebook adding your query and answering the questions. 

### **DataSet**

### <span style="font-size: 14px; font-family: -apple-system, BlinkMacSystemFont, sans-serif;">We will be using the Goodbooks-10k dataset from the exercises in the prep work.&nbsp; Feel free reference your previous notebooks.</span>

- The dataset can be found here: [goodbooks-10k](https://www.kaggle.com/zygmunt/goodbooks-10k)
- You can access the **BooksDB** in the Launchcode server.

### **Business Issue**

You are an employee at a small independet bookstore and you have been tasked to increase sales.  You decide to use the BooksDB to find books and themes to highlight in fun promotions throughout each month / season.  We will be starting with March and then you will have a chance to explore another month of your choosing.  We want to come up with a list of events/ promotions to run each month.  If you are looking for ideas to include in your promotions:

- [https://www.calendarr.com/united-states/calendar-2022/](https://www.calendarr.com/united-states/calendar-2022/)
- [https://www.holidayinsights.com/moreholidays/](https://www.holidayinsights.com/moreholidays/)
    - Click on a month and it will take you to a more detailed page

## Part 1:  March - Women's History Month, Dr Seuss Day(3/2), National Pie Day3/14, St Patrick's Day (3/17), Season - Spring

### **Event 1:** Women's History Month

Goal: Highlight Popular Womans Authors based on ratings from the Good Reads DB.

### **Write a query that returns tag\_id, the number of times each tag\_id is used and the tag\_name.**  

### Use the GROUP BY and HAVING clause to narrow your focus.

Try multiple keywords, ie woman vs female (female ended up giving better results)

In [20]:
-- Solution
SELECT DISTINCT bt.tag_id, t.tag_name, COUNT(t.tag_name) AS tag_count
FROM BooksDB.dbo.book_tags as bt
INNER JOIN BooksDB.dbo.tags as t
ON bt.tag_id = t.tag_id
WHERE t.tag_name LIKE '%female%'
GROUP BY bt.tag_id, t.tag_name
ORDER BY tag_count DESC;

(35 rows affected)

Total execution time: 00:00:00.183

tag_id tag_name tag_count 11643 female-author 997 11644 female-authors 721 11658 female-protagonist 171 11646 female-detective 32 11662 female-writers 18 11642 female 17 11656 female-pov 17 11652 female-lead 14 3514 author-female 12 28834 strong-female-lead 10 11647 female-detectives 9 11650 female-friendship 4 28830 strong-female 4 28833 strong-female-characters 3 11645 female-biographies 2 1741 afghan-female-entrepreneur-under-ta 1 4309 best-adult-female-friendship-books 1 11648 female-doctor 1 11649 female-fiction 1 11651 female-genre-authors 1 11653 female-main-character 1 11654 female-persuasion 1 11655 female-pi 1 11657 female-povs 1 11659 female-serial-killer 1 11660 female-spy 1 11661 female-writer 1 15331 important-female-authors 1 17079 kick-ass-female 1 24455 protagonist-female 1 28831 strong-female-character 1 28832 strong-female-character-s 1 28835 strong-female-protagonist 1 28836 strong-female-supporting-characters 1 29324 tbr-ax-female 1

### **Which keyword(s) and tag(s) would you like to explore more?**

17079 kick-ass-female,  28831 strong-female-character

### **Write a query to return authors, titles, ratings and** **selected tag(s)\_id.**

In [38]:
-- Solution
SELECT b.title, b.average_rating, b.authors, b.book_id, t.tag_id, t.tag_name, COUNT(*) AS times_used
FROM BooksDB.dbo.books AS b
FULL JOIN BooksDB.dbo.tags AS t
ON b.book_id = t.tag_id
WHERE t.tag_id = 11643
GROUP BY b.title, b.average_rating, b.authors, b.book_id, t.tag_id, t.tag_name
--HAVING COUNT(*) > 100
ORDER BY times_used;


(1 row affected)

Total execution time: 00:00:00.043

title average_rating authors book_id tag_id tag_name times_used NULL NULL NULL NULL 11643 female-author 1

### **Record your thoughts about why you wrote the query the way you did.**

# Part 2: Choose Another Month

Choose another month and plan at least 2 events / promotions and answer the following questions.

Which month did you choose?  

What 2 events / promotions are you highlighting?

> 1. November

> 2. Pumpkin Spiced Latte (Coffee)  /  Pumpkins

# Part 3: Summarize your Work

For each event write at least one query that joins any two tables in the BooksDB to support your choice and record you thoughts as to why you used the paticlular query. At least one of your queries needs to include a HAVING clause.

In [39]:
-- Event 1 Query
SELECT DISTINCT bt.tag_id, t.tag_name, COUNT(t.tag_name) AS tag_count
FROM BooksDB.dbo.book_tags as bt
INNER JOIN BooksDB.dbo.tags as t
ON bt.tag_id = t.tag_id
WHERE t.tag_name LIKE '%coffee%'
GROUP BY bt.tag_id, t.tag_name
ORDER BY tag_count DESC;

(11 rows affected)

Total execution time: 00:00:00.171

tag_id tag_name tag_count 7612 coffee-table 30 7614 coffee-table-books 23 7613 coffee-table-book 4 7615 coffee-table-reading 4 7610 coffee 2 7611 coffee-cakes-and-sweets-reading 1 7616 coffeetable 1 8152 cookbooks-and-coffeetables 1 8228 cooking-crafts-coffee-tables 1 20796 my-coffeetable-books 1 21704 non-fiction-coffee-table 1

In [45]:
SELECT b.title, b.average_rating, b.authors, b.book_id, t.tag_id, t.tag_name, COUNT(*) AS times_used
FROM BooksDB.dbo.books AS b
FULL JOIN BooksDB.dbo.tags AS t
ON b.book_id = t.tag_id
WHERE t.tag_id = 7610
GROUP BY b.title, b.average_rating, b.authors, b.book_id, t.tag_id, t.tag_name
--HAVING COUNT(*) > 100
ORDER BY times_used;


(1 row affected)

Total execution time: 00:00:00.049

title average_rating authors book_id tag_id tag_name times_used NULL NULL NULL NULL 7610 coffee 1

### **Summary of event 1:**

Double Click to edit.

In [41]:
-- Event 2 Query
SELECT DISTINCT bt.tag_id, t.tag_name, COUNT(t.tag_name) AS tag_count
FROM BooksDB.dbo.book_tags as bt
INNER JOIN BooksDB.dbo.tags as t
ON bt.tag_id = t.tag_id
WHERE t.tag_name LIKE '%pumpkin%'
GROUP BY bt.tag_id, t.tag_name
ORDER BY tag_count DESC;

(1 row affected)

Total execution time: 00:00:00.153

tag_id tag_name tag_count 24652 pumpkins 1

In [44]:
SELECT b.title, b.average_rating, b.authors, b.book_id, t.tag_id, t.tag_name
INNER JOIN BooksDB.dbo.tags AS t
ON b.book_id = t.tag_id
WHERE t.tag_id = 24652
GROUP BY b.title, b.average_rating, b.authors, b.book_id, t.tag_id, t.tag_name
HAVING COUNT(*) > 100
ORDER BY b.title;


: Msg 156, Level 15, State 1, Line 2
Incorrect syntax near the keyword 'INNER'.

Total execution time: 00:00:00.032

### **Summary of event 2:**

Double Click to edit.